### PR TITLE
Handle providers properly in multiplex log handler

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/apple/swift-log/issues/253

The handling is a bit tricky, but we uphold reasonable semantics now.
We could cache the calculation a bit more perhaps rather than each time in the `get{}` but this has the right semantics at least.

### CODE REVIEW

**Comment by ktoso** [comment]
Please have a look @slashmo

File: Sources/Logging/Logging.swift [hunk 1]
```diff
@@ -1050,6 +1052,8 @@ extension Logger {
 public struct MultiplexLogHandler: LogHandler {
     private var handlers: [LogHandler]
     private var effectiveLogLevel: Logger.Level
+    /// This provider is in addition to any providers the handlers this multiplexes to may have.
```
Comment (slashmo) [inline Sources/Logging/Logging.swift:1055]:
```suggestion
    /// This metadata provider runs after all metadata providers of the multiplexed handlers.
```

**Review by slashmo — APPROVED** [review]
LGTM. In my opinion, the semantics make sense.

**Review by tomerd — APPROVED** [review]

File: Sources/Logging/Logging.swift [hunk 2]
```diff
@@ -1072,6 +1083,43 @@ public struct MultiplexLogHandler: LogHandler {
...
+                    }
+                } else if let multiplexHandlerProvider = self._metadataProvider {
+                    return multiplexHandlerProvider
+                } else {
+                    return nil
+                }
+            } else {
+                var providers: [Logger.MetadataProvider] = []
+                let hasSpecifiedAdditionalMetadataProvider = (self._metadataProvider != nil ? 1 : 0)
+                providers.reserveCapacity(self.handlers.count + hasSpecifiedAdditionalMetadataProvider)
```
Comment (ktoso) [inline Sources/Logging/Logging.swift:1103]:
```suggestion
                providers.reserveCapacity(self.handlers.count + additionalMetadataProviderCount)
```

**Pull request merged**